File: matlab_reference/N-ret-mod---Source-main/ReadSurfer7.m
function [DataOut,Info] = ReadSurfer7(name)

fr = matlab.io.datastore.DsFileReader(name);
disp('Reading GRD File')

% first 4 bytes for file type
[d,~] = read(fr,4,'OutputType','char');
disp(['FileType - ',d]);

if(strcmp(string(d),"DSRB"))
    disp('File format supported!');
else
    error('File format not supported!');
end

%Read Size of Header Section and Version number
[HeaderSize,~] = read(fr,4,'OutputType','int32');
[HeaderVersion,~] = read(fr,4,'OutputType','int32');

 %Read Section Name (expected: GRID)
 [SectionName,~] = read(fr,4,'OutputType','char');

if SectionName == 'GRID'
disp('Reading Grid Section')
end

 %sec2 should be the integer 72, which is the length of the grid section
 [sec2,~] = read(fr,4,'OutputType','int32');

 %Grid Size
 [sec3,~] = read(fr,8,'OutputType','int32');
 disp(['Grid Size: ',num2str(sec3(1)),' rows and ',num2str(sec3(2)),' columns']);
 GridRows = sec3(1);
 GridColumns = sec3(2);
 
 %Grid Corner Location
 [sec4,~] = read(fr,16,'OutputType','double');
 disp(['xLL: ', num2str(sec4(1)),' ','yLL: ', num2str(sec4(2))]);
 GridxLL = sec4(1);
 GridyLL = sec4(2);
 
 %Grid Resolution
 [sec5,~] = read(fr,16,'OutputType','double');
 disp(['x-size: ', num2str(sec5(1)),'m, ','y-size: ', num2str(sec5(2)),'m']);
 GridResolutionX = sec5(1);
 GridResolutionY = sec5(2);

 %Minimum and Maximum Depths
 [sec6,~] = read(fr,16,'OutputType','double');
 Max_Z = sec6(2);
 Min_Z = sec6(1);

 disp(['Min. Z: ', num2str(sec6(1)),'m, ','Max. Z: ', num2str(sec6(2)),'m']);

 %Rotation of Grid (usually is 0)
 [GridRotation,~] = read(fr,8,'OutputType','double');

 %BlankValue
 [BlankValue,~] = read(fr,8,'OutputType','double');

 %Read Section Name (expected: DATA)
 [SectionName2,~] = read(fr,4,'OutputType','char');

 %Read Section Length in Bytes
 [DataSectionLength,~] = read(fr,4,'OutputType','int32');
 disp(['Data Contains ',num2str(DataSectionLength/1024),' kilobytes']);

 disp(['Reading Data'])
 [Data,~] = read(fr,DataSectionLength,'OutputType','double');
 Data(Data==BlankValue) = NaN;
 Z = reshape(Data,[GridColumns,GridRows])';

 DataOut = Z;

 Info.UTM_X = GridxLL+(0:GridResolutionX:(GridColumns-1)*100);
 Info.UTM_Y = GridyLL+(0:GridResolutionY:(GridRows-1)*100);
 
 Info.BlankValue = BlankValue;

 Info.xLL = GridxLL;
 Info.yLL = GridyLL;

 Info.Cols = GridColumns;
 Info.Rows = GridRows;

 Info.MaxZ = Max_Z;
 Info.MinZ = Min_Z;

 Info.xResolution = GridResolutionX;
 Info.yResolution = GridResolutionY;

 Info.GridRotation = GridRotation;
 disp(['File Loaded Successfully'])
 end

File: ReadSurfer7.py
import struct
import numpy as np

def ReadSurfer7(filename):
    with open(filename, 'rb') as f:
        print("Reading GRD File")

        # --- File type ---
        d = f.read(4).decode('ascii')
        print(f"FileType - {d}")

        if d != "DSRB":
            raise ValueError("File format not supported!")
        else:
            print("File format supported!")

        # --- Header size and version ---
        HeaderSize = struct.unpack('<i', f.read(4))[0]
        HeaderVersion = struct.unpack('<i', f.read(4))[0]

        # --- Section name (expected: GRID) ---
        SectionName = f.read(4).decode('ascii')
        if SectionName == "GRID":
            print("Reading Grid Section")

        # --- Grid section length ---
        sec2 = struct.unpack('<i', f.read(4))[0]

        # --- Grid size (rows, cols) ---
        sec3 = struct.unpack('<2i', f.read(8))
        GridRows, GridColumns = sec3
        print(f"Grid Size: {GridRows} rows and {GridColumns} columns")

        # --- Lower-left corner coordinates ---
        sec4 = struct.unpack('<2d', f.read(16))
        GridxLL, GridyLL = sec4
        print(f"xLL: {GridxLL}, yLL: {GridyLL}")

        # --- Grid resolution ---
        sec5 = struct.unpack('<2d', f.read(16))
        GridResolutionX, GridResolutionY = sec5
        print(f"x-size: {GridResolutionX} m, y-size: {GridResolutionY} m")

        # --- Min/Max Z ---
        sec6 = struct.unpack('<2d', f.read(16))
        Min_Z, Max_Z = sec6
        print(f"Min. Z: {Min_Z} m, Max. Z: {Max_Z} m")

        # --- Grid rotation ---
        GridRotation = struct.unpack('<d', f.read(8))[0]

        # --- Blank value ---
        BlankValue = struct.unpack('<d', f.read(8))[0]

        # --- Section name (expected: DATA) ---
        SectionName2 = f.read(4).decode('ascii')

        # --- Data section length ---
        DataSectionLength = struct.unpack('<i', f.read(4))[0]
        print(f"Data Contains {DataSectionLength / 1024:.2f} kilobytes")

        # --- Data section ---
        print("Reading Data")
        num_points = DataSectionLength // 8  # each double = 8 bytes
        Data = np.fromfile(f, dtype='<d', count=num_points)

        # Replace blanks with NaN
        Data[Data == BlankValue] = np.nan

        # Reshape as per Surfer structure.
        # MATLAB's ReadSurfer7.m does ``reshape(Data,[Cols,Rows])'`` where MATLAB
        # reshape is COLUMN-major (Fortran order). NumPy reshape defaults to
        # row-major (C order), so the literal ``reshape((Cols,Rows)).T`` scrambles
        # the data for any non-square grid (it produced striped, mis-classed
        # output). The C-order reshape to (Rows, Cols) reproduces MATLAB exactly:
        # both give Z[r, c] = Data[r*Cols + c].
        Z = Data.reshape((GridRows, GridColumns))

        # --- Output variables ---
        DataOut = Z

        Info = {
            'UTM_X': GridxLL + np.arange(GridColumns) * GridResolutionX,
            'UTM_Y': GridyLL + np.arange(GridRows) * GridResolutionY,
            'BlankValue': BlankValue,
            'xLL': GridxLL,
            'yLL': GridyLL,
            'Cols': GridColumns,
            'Rows': GridRows,
            'MaxZ': Max_Z,
            'MinZ': Min_Z,
            'xResolution': GridResolutionX,
            'yResolution': GridResolutionY,
            'GridRotation': GridRotation,
        }

        print("File Loaded Successfully")

    return DataOut, Info
